### ISSUE TITLE
EFI: FileProtocol.open does not work for directories

### ISSUE THREAD

**Issue by josephlr** [cloud-hypervisor/rust-hypervisor-firmware#19]
Per the UEFI spec, calling open on a directory path should work, but right now it will fail, as the code [here](https://github.com/intel/rust-hypervisor-firmware/blob/f52c1a13d52fdf5a12a111ffff40876209de6b2e/src/fat.rs#L574) is only setup to return a file.

The solution to this is probably reworking the filesystem code to better handle working with directories.

**Comment by rbradford** [comment]
@josephlr yup, I started to refactor the FAT filesystem code itself to expose the directories themselves (in bf6b9b42fc16895f00417499c8573754d5825fea).

Two options I guess:

1. Expose an `open_directory()` function as a mirror to `open()` (don't know if this would work for the EFI case.)
2. Add an object that could be a `Directory` or a `File` that would embed either, we could have a trait that is the union of the functions implemented on the two structs.

Fixing this would be nice as then we could use EFI for systemd-boot.

**Comment by josephlr** [comment]
> @josephlr yup, I started to refactor the FAT filesystem code itself to expose the directories themselves (in [bf6b9b4](https://github.com/intel/rust-hypervisor-firmware/commit/bf6b9b42fc16895f00417499c8573754d5825fea)).

Awesome, I was going to take a crack at the FAT code, but if you're already refactoring it I can take a look after you're done.
 
> Two options I guess:
> 1. Expose an `open_directory()` function as a mirror to `open()` (don't know if this would work for the EFI case.)
> 1. Add an object that could be a `Directory` or a `File` that would embed either, we could have a trait that is the union of the functions implemented on the two structs.

I think having a common enum that could hold a `File or `Directory`, something like the following in the `fat` module:

```rust
pub struct File<'a> { /* ... */ }
pub struct Directory<'a> { /* ... */ }
pub struct Filesystem<'a> { /* ... */ }

pub enum Node<'a> {
    File(File<'a>),
    Directory(Directory<'a>),
}

impl<'a> Directory<'a> {
    pub fn open(&self, path: &str) -> Result<Node, Error>;
    pub fn entries(&self) -> impl Iterator<Item=Node>;
    pub fn get_info(&self) -> DirInfo;
}

impl<'a> File<'a> {
    pub fn read(&mut self, data: &mut [u8]) -> Result<u32, Error>;
    pub fn seek(&mut self, offset: u32) -> Result<(), Error>;
    pub fn get_size(&self) -> u32;
}

impl<'a> Filesystem<'a> {
    pub fn new(device: &'a SectorRead) -> Result<Filesystem, Error>;
    pub fn root(&self) -> Directory;

    pub fn open(&self, path: &str) -> Result<Node, Error> {
        let trimmed_path = /* remove leading "/" */
        self.root().open(trimmed_path)
    }
}
``` 

Then an `efi::FileProtocol` object would just contain a `Node`, allowing easier operation for the EFI methods.

> Fixing this would be nice as then we could use EFI for systemd-boot.

Ya, that's actually how I ran into this issue!

**Comment by rbradford** [comment]
@josephlr I like your idea to use an enum. Apologies if I was unclear, I *did* some refactoring to this end but it's not something i'm working on right now. If you create a PR around the interface you propose above then i'll almost certainly accept it as i'd love to be able to boot systemd-boot under EFI.

**Comment by retrage** [comment]
This feature is implemented as PR #99.
